Add a section on encoding IBCReceive errors (#885)
CosmWasm/cosmwasm: IBC.md

) -> StdResult<IbcReceiveResponse> { }
```

This is a very special entry point as it has a unique workflow. (Please see the
[Acknowledgement Processing section](#Acknowledgement-Processing) below to
understand it fully).

Also note the different return response here (`IbcReceiveResponse` rather than

> `IbcBasicResponse`)? This is because it has an extra field

Reviewer: ```suggestion
`IbcBasicResponse`). This is because it has an extra field
```

`acknowledgement: Binary`, which must be filled out. All successful message must
return an encoded `Acknowledgement` response in this field, that can be parsed
by the sending chain.

The
[`IbcPacket` structure](https://github.com/CosmWasm/cosmwasm/blob/v0.14.0-beta4/packages/std/src/ibc.rs#L129-L146)
contains all information needed to process the receipt. You can generally ignore
timeout (this entry point is only called if it hasn't yet timed out) and
sequence (which is used by the IBC framework to avoid duplicates). I generally

> use `dest.channel_id` like `info.sender` to authenticate the packet, and parse

Author: Yes, this info comes from the ibc module after it is validated. Only when everything is verified does the contract receive the various `ibc_*` packets. I should make this more clear. `packet IbcPacket` is trusted data, just like `info MessageInfo`.

acknowledgement message (that can be parsed by the sending chain as an error).

The atomicity issue was first
[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
and refined into
[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
which was
[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
future-proof interface for contracts, we will use an approach inspired by that
work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
that implements this.

After quite some
[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
we identified a number of issues implementing this idea with the CosmWasm model:

1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
   arbitrary binary encoded message in the error variant.
2. If we return `messages` to be dispatched, they may fail with an error, which
   is returned directly to the x/wasm handler. Obviously an error from the bank

>    handler will not be in the proper IBC Acknowledgement format.

Reviewer: Is this paragraph about messages or sub-messages? Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

Author: > Because for messages the transaction is terminated and reverted, such that no acknowledgement can be created, no matter if the format is known.

Oh, this will change for this one entry point. That is the crux of this proposal. An error from a message will be handled like an error from `ibc_receive_packet`, which means it will be sent to the encoder.

Please let me know how I can express this clearer.

Reviewer: I would not do this. This changes fundamental principles of how messages work. What's preventing `ibc_receive_packet` implementors to use sub messages instead of messages?

3. We also realized that in some cases, the original `ibc_packet_receive` will
   not even have the information needed to format the success response, so we
   will need some delayed response-builder in this case also.

> eg. In the ICS27 spec, they define a packet to register a new account, the

Reviewer: ```suggestion
Eg. in the ICS27 spec, they define a packet to register a new account, the
```

Reviewer: Can we find a nicer way of saying "they"? E.g. "the authors", "it is defined"

acknowledgement of that packet should contain the address of the new account on
the receiving chain. If we implemented this in CosmWasm, the contract would have
to return a submessage of `WasmMsg::Instantiate{}` and only get the proper
address for the response in `reply`. Thus, we need to have some way to fully
build the success acknowledgement after the fact as well.

With this in mind, we designed the following workflow for the `x/wasm` runtime:

1. Wrap the storage with a "local cache", so we don't immediately write out
   changes
2. Call `ibc_receive_packet`, get either `Err(String)` or
   `Ok(Response + acknowledgement)`. If `Err`, jump to 4.
3. Execute any `submessages` and `messages` returned with this response. If any

>    return `Err`, jump to 4.

Reviewer: Do we really want to change how messages behave? I'd avoid that

Reviewer 2: This is also not clear to me. Why do we need step 4 for messages? Submessages are available instead.

Author: `on_packet_receive` may return an error (revert contract state)

`OnPacketReceive` in `x/wasm` should *never* return an error, but rather revert the contract state and return binary data to the sender to explain it failed.

If my contract sends tokens, but for some reason that fails (insufficient funds), it should not abort the tx (so the same packet is posted again and again), but make no state change and return an error acknowledgement.

I suggest reading through the 3 linked cosmos-sdk/ibc-go issues/PRs and all the comments for full context. (I know it is long, but this is the approach they are going for in the sdk and what I want to support in CosmWasm)

Author: In go terms, the `ibc_encode_acknowledgement` works like a `defer()` on  the result

Author: If you understand what I am trying to explain I would be grateful for a better way to express it

4. `receiveResult` is `Err(String)` if any of the above errored, or `Ok(Binary)`
   with the original acknowledgement bytes if we have the success case.
5. Call `ibc_encode_acknowledgement` with the `receiveResult`. The return value

>    is the acknowledgement packet to send to the calling chain. This should never

Reviewer 2: What is the DB transaction scope of this call? Is it the same as `ibc_receive_packet` so that it would be reverted on the error before or a new one that is independent?
I don't have a concrete use case but it feels useful to persist data with a failure ACK as well. If we want this then an error return value makes sense. Though the process would end without an ACK to the packet sender.

Author: This is not my invention, I am following the ibc design (which will come out in 0.43 or 0.44). Check the links at the top of https://github.com/CosmWasm/cosmwasm/blob/on-packet-recv-error-handling/IBC.md#acknowledgement-processing

Especially read the issue description here: https://github.com/cosmos/ibc-go/issues/91 (skim the comments if you wish, the description has the intention, the rest is discussing details). And if you want to dig into the code to understand one part, you can see https://github.com/cosmos/ibc-go/pull/107 (but that is not necessary to get the design)

   error.
6. If `ReceiveResult` is `Ok`, commit this "storage cache". If it is `Err`,

>    revert those changes.

Reviewer 2: 🤔 Do I understand this correct that this step let the contract handle the error case on messages and control the TX scope? The process I was assuming would revert the TX on error in 3) but can persist new state in 5).

Author: The future IBC design (linked above, here again: https://github.com/cosmos/ibc-go/issues/91) explicitly makes the entire app execution (x/wasm) atomic - to be rolled back when app error. But not touching the IBC part of the transaction and successfully committing a "error acknowledgement" in the IBC Keeper.

I feel it is pointless to design something other than what they are doing, unless you want to re-open that issue in the ibc repo. As in the mid-term, we will just use those semantics in `x/wasm` and all other modules will use them as well.

7. Pass the acknowledgement bytes to the IBC router to return

```rust
#[entry_point]
pub fn ibc_encode_acknowledgement(
    // maybe we pass some data from reply via storage
    deps: DepsMut,
    env: Env,
    // the original packet we got if we need some data from that
    packet: IbcPacket,

>     // Either Ok(ibc_receive_packet acknowledgement) or Err(failure message)
>     receiveResult: Result<Binary, String>,

Reviewer: I'd split those to allow accessing the "receive data" in the error case as well:

```suggestion
    // Data returned by `ibc_receive_packet`
    receive_data: Binary,
    // Either Ok(()) or Err(failure message)
    processing_result: Result<(), String>,
```

Author: Okay, so `ibc_receive_packet` may return `acknowledgement: Binary` or may error.

If it succeeded but a dispatched (sub-)message errored, then we have both an error (eg. from BankMsg) and an acknowledgement (from the original call).

We end up with:

```
receive_data: Option<Binary>,
processing_result: Result<(), String>,
```

Does that make sense?

Reviewer: I would not call the ack encoder/creator entry point if `ibc_receive_packet` errors but instead fail the transaction execution (same discussions as with messages).

Then `receive_data` is always set (to whatever `ibc_receive_packet` returned).

Author: Continuing on discord. I have clearly failed to explain the IBC architecture design that we fit into.

) -> StdResult<Binary> { }
```

The function receives the original packet as well as the error string returned,
and is responsible for creating a proper binary-encoded "error acknowledgement"
that can be sent to the calling contract. For example, it could protobuf-encode
the `Acknowledgement message` defined below. `DepsMut`, `Env` and `IbcPacket`
are passed just in case more context is needed for encoding, but likely unused.
This function should never return an error (even if the IbcPacket was
malformed), but we allow it to return one if there is some pathological state

> (rather than panicking in the contract,`StdResult::Err` returns useful

Reviewer: ```suggestion
(rather than panicking in the contract, `StdResult::Err` returns useful
```

properly commit an acknowledgement with encoded error. In other words, this "IBC
Handler" will error and revert, but the "IBC Router" must succeed and commit an
acknowledgement message (that can be parsed by the sending chain as an error).

The atomicity issue was first
[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
and refined into
[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
which was
[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
future-proof interface for contracts, we will use an approach inspired by that
work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
that implements this.

After quite some
[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
we struggled to map this idea to the CosmWasm model. However, we also discovered
a deep similarity between these requirements and the
[submessage semantics](./SEMANTICS.md#submessages). It just requires some

> careful coding on the contract developer's side to new throw errors. This

Reviewer: Is this what you want to say?

```suggestion
careful coding on the contract developer's side to not throw errors. This
```

> produced 3 suggests on how to handle errors and rollbacks _inside

Reviewer: ```suggestion
produced 3 suggestions on how to handle errors and rollbacks _inside
```

`ibc_packet_receive`_

1. If the message doesn't modify any state directly, you can simply put the
   logic in a closure, and capture errors, converting them into error
   acknowledgements. This would look something like the
   [main dispatch loop in `ibc-reflect`](https://github.com/CosmWasm/cosmwasm/blob/cd784cd1148ee395574f3e564f102d0d7b5adcc3/contracts/ibc-reflect/src/contract.rs#L217-L248):

> ```rust

Reviewer: You can indent it such that the whole code block belongs to 1., not below 1.:

```suggestion
   ```rust
```

    (|| {
        // which local channel did this packet come on
        let caller = packet.dest.channel_id;
        let msg: PacketMsg = from_slice(&packet.data)?;
        match msg {
            PacketMsg::Dispatch { msgs } => receive_dispatch(deps, caller, msgs),
            PacketMsg::WhoAmI {} => receive_who_am_i(deps, caller),
            PacketMsg::Balances {} => receive_balances(deps, caller),
        }
    })()
    .or_else(|e| {
        // we try to capture all app-level errors and convert them into
        // acknowledgement packets that contain an error code.
        let acknowledgement = encode_ibc_error(format!("invalid packet: {}", e));
        Ok(IbcReceiveResponse {
            acknowledgement,
            submessages: vec![],
            messages: vec![],
            attributes: vec![],
        })
    })

> ```

Reviewer: We could even enforece this by changing the signature of the entry point to `IbcReceiveResponse`, i.e. no result around it.

Author: Nice idea. This will push devs to proper usage

2. If we modify state with an external call, we need to wrap it in a
   `submessage` and capture the error. This approach requires we use _exactly
   one_ submessage. If we have multiple, we may commit #1 and rollback #2 (see
   example 3 for that case). The main point is moving `messages` to
   `submessages` and reformating the error in `reply`. Note that if you set the
   `Response.data` field in `reply` it will override the acknowledgement
   returned from the parent call. (See
   [bottom of reply section](./SEMANTICS.md#handling-the-reply)). You can see a
   similar example in how
   [`ibc-reflect` handles `receive_dispatch`](https://github.com/CosmWasm/cosmwasm/blob/eebb9395ccf315320e3f2fcc526ee76788f89174/contracts/ibc-reflect/src/contract.rs#L307-L336).

>    Note how we use a unique reply id for this and use that to catch any

Reviewer: ```suggestion
   Note how we use a unique reply ID for this and use that to catch any
```

   execution failure and return an error acknowledgement instead:

> ```rust

Reviewer: ```suggestion
   ```rust
```

    let reflect_addr = accounts(deps.storage).load(caller.as_bytes())?;

    // let them know we're fine
    let acknowledgement = to_binary(&AcknowledgementMsg::<DispatchResponse>::Ok(()))?;
    // create the message to re-dispatch to the reflect contract
    let reflect_msg = ReflectExecuteMsg::ReflectMsg { msgs };
    let wasm_msg = wasm_execute(reflect_addr, &reflect_msg, vec![])?;

    // we wrap it in a submessage to properly report errors
    let sub_msg = SubMsg {
        id: RECEIVE_DISPATCH_ID,
        msg: wasm_msg.into(),
        gas_limit: None,
        reply_on: ReplyOn::Error,
    };

    Ok(IbcReceiveResponse {
        acknowledgement,
        submessages: vec![sub_msg],
        messages: vec![],
        attributes: vec![attr("action", "receive_dispatch")],
    })
}

#[entry_point]
pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> StdResult<Response> {
   match (reply.id, reply.result) {
      (RECEIVE_DISPATCH_ID, ContractResult::Err(err)) => Ok(Response {
         data: Some(encode_ibc_error(err)),
         ..Response::default()
      }),
      (INIT_CALLBACK_ID, ContractResult::Ok(response)) => handle_init_callback(deps, response),

>       _ => Err(StdError::generic_err("invalid reply id")),

Reviewer: ```suggestion
      _ => Err(StdError::generic_err("invalid reply id or result")),
```

Although it suggests this is a Protobuf object, the ICS spec doesn't define
whether to encode it as JSON or Protobuf. In the ICS20 implementation, this is

> JSON encoded when returned from a contract. In ICS27, they are discussing using

Reviewer: ```suggestion
JSON encoded when returned from a contract. In ICS27, the authors are discussing using
```

Although it suggests this is a Protobuf object, the ICS spec doesn't define
whether to encode it as JSON or Protobuf. In the ICS20 implementation, this is
JSON encoded when returned from a contract. In ICS27, the authors are discussing
using a Protobuf-encoded form of this structure.

Note that it leaves the actual success response as app-specific bytes where you
can place anything, but does provide a standard way for an observer to check
success-or-error. If you are designing a new protocol, I encourage you to use

> this struct in either of the encodings as the acknowledgement envelope.

Reviewer 2: 👍  good to have this as recommendation
